packages/core/conductor-tasks/common/lockS3File.ts

```diff
@@ -0,0 +1,83 @@
+import type { ConductorWorker } from "@io-orkes/conductor-javascript"
+import completed from "@moj-bichard7/common/conductor/helpers/completed"
+import failed from "@moj-bichard7/common/conductor/helpers/failed"
+import failedTerminal from "@moj-bichard7/common/conductor/helpers/failedTerminal"
+import inputDataValidator from "@moj-bichard7/common/conductor/middleware/inputDataValidator"
+import type Task from "@moj-bichard7/common/conductor/types/Task"
+import createS3Config from "@moj-bichard7/common/s3/createS3Config"
+import readTagsFromS3 from "@moj-bichard7/common/s3/readTagsFromS3"
+import writeTagsToS3 from "@moj-bichard7/common/s3/writeTagsToS3"
+import { isError } from "@moj-bichard7/common/types/Result"
+import { z } from "zod"
+
+const s3Config = createS3Config()
+
+const inputDataSchema = z.object({
+  fileName: z.string(),
+  bucketId: z.string(),
+  workflowId: z.string()
+})
+type InputData = z.infer<typeof inputDataSchema>
+
+const { TASK_DATA_BUCKET_NAME } = process.env
+if (!TASK_DATA_BUCKET_NAME) {
+  throw Error("TASK_DATA_BUCKET_NAME environment variable is required")
+}
+
+const buckets: Record<string, string> = {
+  "task-data": TASK_DATA_BUCKET_NAME
+}
+
+const lockKey = "lockedByWorkstream"
+
+const deleteS3File: ConductorWorker = {
+  taskDefName: "lock_s3_file",
+  execute: inputDataValidator(inputDataSchema, async (task: Task<InputData>) => {
+    // This provides basic locking to try to make sure the workflow doesn't get triggered twice.
+    // This implementation is not race-condition proof. However, the specific case we have
+    // with duplicate workflows separated by a period of time will be prevented by this.
+    // Currently there's a window of 2 - 5 seconds where multiple workflows could be started
+    // for the same message - this reduces the window to a few milliseconds
+    const { fileName, bucketId, workflowId } = task.inputData
+
+    const bucket = buckets[bucketId]
+    if (!bucket) {
+      return failedTerminal(`Bucket ID was not found: ${bucketId}`)
+    }
+
+    const tags = await readTagsFromS3(fileName, bucket, s3Config)
+    if (isError(tags)) {
+      // File has already been removed
+      if (tags.name === "NoSuchKey") {
+        return completed({ lockState: "failure" }, "S3 File already deleted")
+      }
+
+      // Error reading the tags
+      return failed(tags.message)
+    }
+
+    if (tags[lockKey]) {
+      // File has already been locked
+      return completed({ lockState: "failure" }, "S3 File already locked")
+    }
+
+    const result = await writeTagsToS3(fileName, bucket, { [lockKey]: workflowId }, s3Config)
+    if (isError(result)) {
+      return failed(result.message)
+    }
+
+    const newTags = await readTagsFromS3(fileName, bucket, s3Config)
+    if (isError(newTags)) {
+      return failed(tags.message)
```

Comment:
```suggestion
      return failed(newTags.message)
```

Should this be using `newTags` instead?

Reply:
👍